pytorch_lightning/core/memory.py

```diff
@@ -137,6 +137,8 @@ class ModelSummary(object):
              - `top` (default): only the top-level modules will be recorded (the children of the root module)
              - `full`: summarizes all layers and their submodules in the root module
 
+        max_depth: Maximum depth of modules to show when mode="full"
```

Comment:
If we really want `max_depth`, we would have to drop `mode` I'm pretty sure because two arguments that have overlapping functionality is not good to maintain, i.e., `mode="top"` is the same as `max_depth=1` and `mode="full"` is the same as `max_depth = -1`.  

I propose the following: Drop `mode` on  `ModelSummary` completely. Map `Trainer.weights_summary=full|top|None` and `Lightning.summarize` to the appropriate value of `max_depth`.

Reply:
this would also avoid your error handling below

Reply:
Yes, I agree, although I did not want to make a breaking change. Your proposal seems to resolve things nicely if I understand correctly. I just have some questions regarding the implementation:

* Should I leave the `ModelSummary.MODES`? this is used in checks in `Lightning.summarize()`, `Trainer.weights_summary` and in the pytests i.e., `if mode in ModelSummary.MODES:`. Or do I just replace the checks with: `if mode in ("top", "full"):`)? or move the variables elsewhere?
* To make full use of the max_depth functionality, the user would be required to use the `ModelSummary` class directly, leaving the old interface with `mode` in the `Trainer` and `Lightning.summarize` right?

Reply:
@awaelchli How about having an argument that could support both?

Reply:
Forgot to say thanks for sending a PR and your ideas! Greatly appreciated. 

It would be a change but not a serious one. We would of course not immediately remove it but deprecate the argument. And since we would leave the trainer flag intact and only touch ModelSummary directly, not many users are affected :) 

I suggest we stick to the contents of this PR, with the addition of adding a deprecation message. Then, in a following PR re-route the old mode argument to max_depth (and support -1) and update tests. Until deprecation is done, `ModelSummary.MODES` can stay or be hardcoded and eventually be removed.

> To make full use of the max_depth functionality, the user would be required to use the ModelSummary class directly, leaving the old interface with mode in the Trainer and Lightning.summarize right?

That's actually how many features in Lightning are controlled. A simple on/off switch in the Trainer for convenience and for customization we would interact with the feature instances directly, e.g., loggers callbacks profilers etc. So I think that pattern for model summary fits well. `max_depth` could still remain an argument to summarize() as you have it now.

Of course, other reviewers may also have some input.

Reply:
I've already added the mapping from `mode` (in Trainer / PL.summarize) to max_depth to the `ModelSummary` and updated the tests accordingly.

The `PL.summarize()` can work with `mode` or `max_depth` parameters, but will raise a warning if only `mode` is passed.

Let me know if you agree.